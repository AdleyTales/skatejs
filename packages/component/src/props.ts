import { PropType } from './types';

const any: PropType = {
  default: (elem, name, oldValue) => null,
  defined: (ctor, name) => {},
  deserialize: (elem, name, oldValue, newValue) => newValue,
  get: (elem, name, oldValue) => oldValue,
  serialize: (elem, name, oldValue, newValue) => newValue,
  set: (elem, name, oldValue, newValue) => newValue,
  source: propName => propName.toLowerCase(),
  target: propName => {}
};

const array: PropType = {
  ...any,
  default: () => [],
  deserialize: (elem, name, oldValue, newValue) => JSON.parse(newValue),
  serialize: (elem, name, oldValue, newValue) => JSON.stringify(newValue)
};

const boolean: PropType = {
  ...any,
  deserialize: (elem, name, oldValue, newValue): boolean => newValue != null,
  serialize: (elem, name, oldValue, newValue) => (newValue ? '' : null)
};

const event: PropType = {
  ...any,
  set(elem, name, oldValue, newValue) {
    // TODO see if we can deserialize to a standard onclick prop so that we
    // can support in-attribute handlers.
    const eventName = this.getEventName(name);
    if (oldValue) {
      elem.removeEventListener(eventName, oldValue);
    }
    if (newValue) {
      elem.addEventListener(eventName, newValue);
    }
    return newValue;
  },
  getEventName(name: string): string {
    return name;
  }
};

const number: PropType = {
  ...any,
  default: () => 0,
  deserialize: (elem, name, oldValue, newValue): number =>
    newValue == null ? 0 : Number(newValue),
  serialize: (elem, name, oldValue, newValue) =>
    newValue == null ? null : String(Number(newValue))
};

const object: PropType = {
  ...array,
  default: () => ({})
};

const string: PropType = {
  ...any,
  default: () => '',
  deserialize: (elem, name, oldValue, newValue) => newValue,
  serialize: (elem, name, oldValue, newValue) =>
    newValue == null ? null : String(newValue)
};

export const props = {
  any,
  array,
  boolean,
  event,
  number,
  object,
  string
};
