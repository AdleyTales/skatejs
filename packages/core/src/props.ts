import { PropType } from './types';

const any: PropType = {
  defined: (elem, name) => {},
  default: (elem, name) => null,
  deserialize: (elem, name, oldValue, newValue) => newValue,
  get: (elem, name, value) => value,
  serialize: (elem, name, oldValue, newValue) => newValue,
  set: (elem, name, oldValue, newValue) => newValue,
  source: propName => propName.toLowerCase(),
  target: propName => {}
};

const object: PropType = {
  ...any,
  deserialize: (elem, name, oldValue, newValue) => JSON.parse(newValue),
  serialize: (elem, name, oldValue, newValue) => JSON.stringify(newValue)
};

const boolean: PropType = {
  ...any,
  deserialize: (elem, name, oldValue, newValue): boolean => newValue != null,
  serialize: (elem, name, oldValue, newValue) => (newValue ? '' : null)
};

const event: PropType = {
  ...any,
  set(elem, name, oldValue, newValue) {
    // TODO see if we can deserialize to a standard onclick prop so that we
    // can support in-attribute handlers.
    const eventName = this.getEventName(name);
    if (oldValue) {
      elem.removeEventListener(eventName, oldValue);
    }
    if (newValue) {
      elem.addEventListener(eventName, newValue);
    }
    return newValue;
  },
  getEventName(name: string): string {
    return name;
  }
};

const number: PropType = {
  ...any,
  deserialize: (elem, name, oldValue, newValue): number =>
    newValue == null ? 0 : Number(newValue),
  serialize: (elem, name, oldValue, newValue) =>
    newValue == null ? null : String(Number(newValue))
};

const array: PropType = object;

const string: PropType = {
  ...any,
  deserialize: (elem, name, oldValue, newValue) => newValue,
  serialize: (elem, name, oldValue, newValue) =>
    newValue == null ? null : String(newValue)
};

export default {
  any,
  array,
  boolean,
  event,
  number,
  object,
  string
};
