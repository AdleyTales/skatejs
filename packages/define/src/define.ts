import name from './name';

export interface Constructor<T> {
  new (): T;
  is?: string;
  name: string;
}

export default function<T>(ctor: Constructor<T>): Constructor<T> {
  if (!ctor.is) {
    ctor.is = name(ctor.name);
    customElements.define(ctor.is, ctor);
  }
  return ctor;
}
