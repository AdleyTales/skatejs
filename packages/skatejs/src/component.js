// @flow

import { withComponent } from './with-component.js';

// This is so the transform can patch this class and everything extending it
// doesn't need to worry about a shim for native / polyfilled custom elements.
class PatchedHTMLElement extends HTMLElement {}

// This is merely convenience that enables:
//
// - Not having to patch HTMLElement (shim or transform).
// - Doing: withComponent(HTMLElement).
export const Component = withComponent(PatchedHTMLElement);
