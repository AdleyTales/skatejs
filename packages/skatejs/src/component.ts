import { CustomElementConstructor } from './types';
import { observe } from './observe';
import { shadow } from './shadow';

function delay(fn) {
  if (typeof global.Promise === 'function') {
    // @ts-ignore - Promise.resove() indeed does exist.
    global.Promise.resolve().then(fn);
  } else {
    setTimeout(fn);
  }
}

export function component(
  Base: CustomElementConstructor
): CustomElementConstructor {
  return class extends Base {
    ['constructor']: CustomElementConstructor;

    _boundRender: () => any;
    _prevProps: {} = {};
    _prevState: {} = {};
    _props: {} = {};
    _state: {} = {};
    _updating: boolean = false;
    childrenUpdated?();

    static _attrToPropMap = {};
    static _propToAttrMap = {};

    // This is invoked once when the element is defined. Due to this, we must
    // link props to attrs here as opposed to doing it in the constructor.
    static get observedAttributes(): Array<string> {
      const { attrs, props } = this;
      return Object.keys(props).reduce((observedAttrs, propName) => {
        // We're opinionated about the propName -> propname (attribute)
        // convention.
        const attrName = propName.toLowerCase();

        // Don't set props from attributes if there is no handler. This also
        // means we don't need to observe them at all.
        if (this.props[propName]) {
          this._attrToPropMap[attrName] = propName;
          this._propToAttrMap[propName] = attrName;
          observedAttrs = observedAttrs.concat(attrName);
        }

        // Defines a property to ensure it triggers an updated and syncs the
        // corresponding attribute.
        Object.defineProperty(this.prototype, propName, {
          configurable: true,
          get() {
            return this._props[propName];
          },
          set(propValue) {
            this._props[propName] = propValue;
            if (propName in attrs) {
              const attrName = this._propToAttrsMap[propName];
              const attrValue = attrs[propName](propValue);
              if (attrValue == null) {
                this.removeAttribute(attrName);
              } else {
                this.setAttribute(attrName, attrValue);
              }
            }
            this.triggerUpdate();
          }
        });

        return observedAttrs;
      }, []);
    }

    static get attrs() {
      return this.props;
    }

    static get props() {
      return Object.keys(this.prototype).reduce((o, k) => {
        o[k] = v => v;
        return o;
      }, {});
    }

    get props(): {} {
      return this._props;
    }

    set props(props: {}) {
      this._props = props;
    }

    get renderRoot() {
      return super.renderRoot || shadow(this);
    }

    get state(): {} {
      return this._state;
    }

    set state(state: {}) {
      this._state = state;
      this._triggerUpdate();
    }

    attributeChangedCallback(
      name: string,
      oldValue: string | null,
      newValue: string | null
    ) {
      const { _attrToPropMap, props } = this.constructor;

      if (super.attributeChangedCallback) {
        super.attributeChangedCallback(name, oldValue, newValue);
      }

      const propertyName = _attrToPropMap[name];
      const propertyFunc = props[propertyName];

      if (propertyFunc) {
        this._props[propertyName] = propertyFunc(newValue);
        this._triggerUpdate();
      }
    }

    connectedCallback() {
      if (super.connectedCallback) {
        super.connectedCallback();
      }

      if (this.childrenUpdated) {
        observe(this, this.childrenUpdated.bind(this));
      }

      this._triggerUpdate();
    }

    renderer(root, html) {
      if (super.renderer) {
        super.renderer(root, html);
      } else {
        root.innerHTML = html();
      }
    }

    shouldUpdate(props: {}, state: {}): boolean {
      return true;
    }

    _triggerUpdate(): void {
      if (this._updating) {
        return;
      }

      if (!this._boundRender) {
        this._boundRender = this.render.bind(this);
      }

      // This flag prevents infinite loops if another update is triggered while
      // performing the current update.
      this._updating = true;

      // We execute the update process at the end of the current microtask so
      // we can debounce any subsequent updates using the _updating flag.
      delay(() => {
        const { _prevProps, _prevState } = this;
        if (this.shouldUpdate(_prevProps, _prevState)) {
          this.renderer(this.renderRoot, this._boundRender);
          if (this.updated) {
            this.updated(_prevProps, _prevState);
          }
        }
        this._prevProps = this.props;
        this._prevState = this.state;
        this._updating = false;
      });
    }
  };
}
