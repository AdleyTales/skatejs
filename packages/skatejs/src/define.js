// @flow

import type { CustomElement } from './types.js';
import { name } from './name.js';

export function define(Ctor: CustomElement): CustomElement {
  customElements.define(Ctor.is || name(), Ctor);
  return Ctor;
}
