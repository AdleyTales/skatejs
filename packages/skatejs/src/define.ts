import { CustomElementConstructor } from './types';
import { name } from './name.js';

export function define(
  ctor: CustomElementConstructor
): CustomElementConstructor {
  Object.defineProperty(ctor, 'is', { value: name(ctor.is) });
  customElements.define(ctor.is, ctor);
  return ctor;
}
