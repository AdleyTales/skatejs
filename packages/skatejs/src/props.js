// @flow

const { parse } = JSON;

export const props = {
  any: (v: any) => v,
  array: parse,
  boolean: (v: string | null) => v === '',
  number: Number,
  object: parse,
  string: String
};
