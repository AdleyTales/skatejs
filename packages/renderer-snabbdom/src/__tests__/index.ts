import { wait } from '@skatejs/bore';
import define from '@skatejs/define';
import h from 'snabbdom/h';
import component from '..';

const Component = component([
  require('snabbdom/modules/attributes').default,
  require('snabbdom/modules/eventlisteners').default,
  require('snabbdom/modules/class').default,
  require('snabbdom/modules/props').default,
  require('snabbdom/modules/style').default,
  require('snabbdom/modules/dataset').default
]);

const Test = define(
  class extends Component {
    name: string = 'World';
    static props = { name: String };
    render() {
      return h('span', `Hello, ${this.name}!`);
    }
  }
);

function testContent(text) {
  return `<span>Hello, ${text}!</span>`;
}

test('renders', async () => {
  const el = new Test();
  await wait();
  expect(el.shadowRoot.innerHTML).toEqual('');

  document.body.appendChild(el);
  await wait();
  expect(el.shadowRoot.innerHTML).toEqual(testContent('World'));

  el.name = 'Bob';
  await wait();
  expect(el.shadowRoot.innerHTML).toEqual(testContent('Bob'));

  document.body.removeChild(el);
  await wait();
  expect(el.shadowRoot.innerHTML).toEqual('');
});
