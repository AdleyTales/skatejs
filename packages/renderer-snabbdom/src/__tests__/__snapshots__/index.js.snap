// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanup 1`] = `""`;

exports[`renders 1`] = `""`;

exports[`renders 2`] = `"<div>Hello, World!</div>"`;

exports[`renders 3`] = `"<div>Hello, Bob!</div>"`;
